pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,425 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+				return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+					capacity, aggCapacity, minCapacity, maxCapacity)
+			}
+			return scaleUpByInteger(replicas, capacity, count,
+				aggCapacity, availableCapacity, maxCapacity, buffer)
+		case availableCapacity > buffer: // Scale Down
+			if limited && scale == 1 { // Case where we want to scale down but we're already limited by MinCapacity
+				return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+					capacity, aggCapacity, minCapacity, maxCapacity)
+			}
+			return scaleDownByInteger(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				availableCapacity, aggCount, aggCapacity, minCapacity, buffer)
+		}
+	}
+
+	// Desired replicas based on BufferSize specified as a percent (i.e. 5%)
+	bufferPercent, err := intstr.GetValueFromIntOrPercent(&bufferSize, 100, isCounter)
+	if err != nil {
+		return 0, false, err
+	}
+	// The desired total capacity across the fleet (see applyBufferPolicy for explanation)
+	desiredCapacity := float64(aggCount*100) / float64(100-bufferPercent)
+
+	switch roundedDesiredCapacity := int64(math.Ceil(desiredCapacity)); {
+	case roundedDesiredCapacity == aggCapacity:
+		if limited {
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return replicas, false, nil
+
+	case roundedDesiredCapacity > aggCapacity: // Scale up
+		if limited && scale == -1 { // Case where we want to scale up but we're already limited by MaxCapacity
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return scaleUpByPercent(replicas, count, aggCount, capacity, aggCapacity, maxCapacity,
+			desiredCapacity, float64(bufferPercent))
+
+	case roundedDesiredCapacity < aggCapacity: // Scale down
+		if limited && scale == 1 { // Case where we want to scale down but we're already limited by MinCapacity
+			return scaleLimited(scale, f, gameServerLister, nodeCounts, key, isCounter, replicas,
+				capacity, aggCapacity, minCapacity, maxCapacity)
+		}
+		return scaleDownByPercent(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+			aggCount, aggCapacity, minCapacity, float64(bufferPercent))
+	}
+
+	if isCounter {
+		return 0, false, errors.Errorf("unable to apply CounterPolicy %v", c)
+	}
+	return 0, false, errors.Errorf("unable to apply ListPolicy %v", l)
+}
+
+// getSortedGameServers returns the list of Game Servers for the Fleet in the order in which the
+// Game Servers would be deleted.
+func getSortedGameServers(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount) ([]*agonesv1.GameServer, error) {
+	// TODO: Should we handle this differently for strategy Distributed?
+	gss := f.GetGameServerSpec()
```

Comment:
Should we handle this differently for strategy Distributed?

Reply:
Distributed sorting should be across all the nodes, rather than having counts/lists be the tie-breaker within a node, as it is for Packed -- so I think the answer is yes?

Reply:
I guess I should have asked a different question: This function calls SortGameServersByStrategy which sorts by either Packed or Distributed strategies. In the case of Packed I am fairly confident that the list of gameservers in order of deletion will be the same as when the SortGameServersByStrategy is later called if there are actually gameservers that can be deleted. I'm not confident that we'll get back the same list of game servers when the controller later calls SortGameServersByStrategy when the stragegy is Distributed. Which would mean that the "desiredReplicas" we calculate here would be a different number than if we'd calculated it based on a different ordered list of gameservers.

The general issue is that the FleetAutoscaler Controller is expected "desired replicas" and calls SortGameServersByStrategy later. This isn't the ideal flow for Counts and Lists -- in an ideal scenario we would have the list of game servers in order of deletion first. I'm not sure we want to muck about with the controller that much though?

Reply:
I'll add this in a separate PR just for the sake of keeping this PR from turning into a XXXL.